proposals/NNNN-add-with-methods-to-swift.md

```diff
@@ -0,0 +1,437 @@
+    var copy = self
+    try await transform(&copy)
+    return copy
+  }
+}
+```
+
+Since `extension Any` is not expressible in the surface language today, the actual implementation will require changes to the compiler (which are beyond the scope of a proposal review, although discussion is available in the [implementation PR](https://github.com/apple/swift/pull/67768)).
+
+### `@_disfavoredOverload`
+
+`@_disfavoredOverload` is required to maximize source compatibility. 
+
+Since this proposal adds a new method named `with` to all types / values, this new method potentially conflicts with any existing method named `with`. 
+
+There are several examples that would fail to compile if this method were added without being a disfavored overload. For example:
+
+```swift
+// An existing type that has a `with` function of the same signature
+struct Foo {
+  var bar: String?
+  
+  func with(_ modify: (inout Foo) -> Void) -> Foo {    
+    var copy = self
+    modify(&copy)
+    return copy
+  }
+}
+
+// Without @_disfavoredOverload, produces an error "ambiguous use of 'with'"
+let value = Foo().with { $0.bar = "bar" }
+```
+
+```swift
+// An existing type that has a `with` function of a different signature
+struct Foo {
+  var bar: String?
+  
+  func with(bar: String) -> Foo {    
+    var copy = self
+    copy.bar = bar
+    return copy
+  }
+}
+
+// Without @_disfavoredOverload, produces an error "ambiguous use of 'with'"
+let withFunc = Foo.with
+
+let value = Foo().with(bar: "bar")
+```
+
+Annotating the new `with` methods with `@_disfavoredOverload` allows these examples to continue compiling / functioning as they did before. 
+
+`@_disfavoredOverload` could potentially be removed in Swift 6 mode, where a source break could be more acceptable.
+
+### `@dynamicMemberLookup`
+
+`@dynamicMemberLookup` is currently the overload-of-last-resort, only used if there are no other overloads that match the given name. It is possible to create examples that no longer compile after adding the new `with` method:
+
+```swift
+@dynamicMemberLookup
+struct Foo {
+  var bar: String?
+  
+  subscript(dynamicMember member: String) -> String {
+    "dymamic member \(member)"
+  }
+}
+
+// Today this prints "DYNAMIC MEMBER WITH"
+//
+// With the new `with` method, by default this would produce an error:
+// value of type '((inout Test) -> Void) -> Test' has no member 'uppercased'
+print(Foo().with.uppercased())
+
+// Today this prints "DYNAMIC MEMBER WITH"
+//
+// With the new `with` method, by default this would produce an error: 
+// ambiguous use of 'with'
+let string = Foo().with
+print(string.uppercased())
+```
+
+To prevent this source break, we can make it so that the new `with` overload doesn't suppress the dynamic member subscrpt overload when accessing a dyanmic member named "with". Combining this with `@_disfavoredOverload` allows the above example to continue compiling / functioning exactly as it did before.
```

Comment:
```suggestion
To prevent this source break, we can make it so that the new `with` overload doesn't suppress the dynamic member subscrpt overload when accessing a dynamic member named "with". Combining this with `@_disfavoredOverload` allows the above example to continue compiling / functioning exactly as it did before.
```